src/core/dfn-panel.js

```diff
@@ -0,0 +1,166 @@
+  }
+  if (document.getElementById("dfn-panel")) {
+    return "hide";
+  }
+  return null;
+}
+
+/** @param {HTMLElement} dfn */
+function createPanel(dfn) {
+  const { id } = dfn;
+  const href = `#${id}`;
+  const links = document.querySelectorAll(`a[href="${href}"]`);
+
+  /** @type {HTMLElement} */
+  const panel = hyperHTML`
+    <aside class="dfn-panel" id="dfn-panel">
+      <b><a class="self-link" href="${href}">Permalink</a></b>
+      <b>Referenced in:</b>
+      ${referencesToHTML(id, links)}
+    </aside>
+  `;
+  return document.body.appendChild(panel);
+}
+
+/**
+ * @param {string} id dfn id
+ * @param {NodeListOf<HTMLLinkElement>} links
+ * @returns {HTMLUListElement}
+ */
+function referencesToHTML(id, links) {
+  if (!links.length) {
+    return hyperHTML`<ul><li>Not referenced in this document.</li></ul>`;
+  }
+
+  /** @type {Map<string, string[]>} */
+  const titleToIDs = new Map();
+  links.forEach((link, i) => {
+    const linkID = link.id || `ref-for-${id}-${i + 1}`;
+    if (!link.id) link.id = linkID;
+    const title = getReferenceTitle(link);
+    const ids = titleToIDs.get(title) || titleToIDs.set(title, []).get(title);
+    ids.push(linkID);
+  });
+
+  /**
+   * Takes an "entry" from `titleToIDs`, and converts it to a list that is
+   * easier to render in `listItemToHTML`. The first list item contains title
+   * from `getReferenceTitle`, rest of items contain strings like `(2)`, `(3)`
+   * as title.
+   * @param {[string, string[]]} entry
+   */
+  const toLinkProps = ([title, ids]) => {
+    return [{ title, id: ids[0] }].concat(
+      ids.slice(1).map((id, i) => ({ title: `(${i + 2})`, id }))
+    );
+  };
+
+  /**
+   * @param {[string, string[]]} entry
+   * @returns {HTMLLIElement}
+   */
+  const listItemToHTML = entry =>
+    hyperHTML`<li>${toLinkProps(entry).map(
+      link => hyperHTML`<a href="#${link.id}">${link.title}</a>${" "}`
+    )}</li>`;
+
+  return hyperHTML`<ul>${[...titleToIDs.entries()].map(listItemToHTML)}</ul>`;
+}
+
+/** @param {HTMLAnchorElement} link */
+function getReferenceTitle(link) {
+  const section = link.closest("section");
+  if (!section) return null;
+  const heading = section.querySelector("h1, h2, h3, h4, h5, h6");
+  if (!heading) return null;
+  return norm(heading.textContent);
+}
+
+/**
+ * @param {HTMLElement} dfn
+ * @param {HTMLElement} panel
+ */
+function displayPanel(dfn, panel) {
+  const dfnRect = dfn.getBoundingClientRect();
+  const panelRect = panel.getBoundingClientRect();
+  const panelWidth = panelRect.right - panelRect.left;
+
+  let top = window.scrollY + dfnRect.top;
+  let left = dfnRect.left + dfnRect.width + 5;
+  if (left + panelWidth > document.body.scrollWidth) {
+    // Reposition, because the panel is overflowing
+    left = dfnRect.left - (panelWidth + 5);
+    if (left < 0) {
+      left = dfnRect.left;
+      top = top + dfnRect.height;
+    }
+  }
+
+  // "docked" state can independently set its own position in CSS.
```

Comment:
```suggestion
  // Allows ".docked" rule to override the position, unlike `style.left = left`.
```

IMO better, what do you think?